[wallet] Upgrade path for non-HD wallets to HD

achow101 opened the pull request:
Revival/rebase of #11085

Adds a new command `sethdseed` which allows you to either set or generate a new HD seed to be used. A new keypool can be generated or the original one kept and new keys added to the keypool will come from the new HD seed.

Wallets that are not HD will be upgraded to be version FEATURE_HD_SPLIT when the `sethdseed` RPC command is used.

I have also add some tests for this.

Additionally `-upgradewallet` can now be used to upgrade a wallet from non-HD to HD. When it is used for such an upgrade, the keypool will be regenerated.

dooglus committed:
Add 'sethdseed' RPC to initialize or replace HD seed

achow101 committed:
Test sethdseed

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,71 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            "\nNote that you will need to MAKE A NEW BACKUP of your wallet after setting the HD wallet seed.\n"
+            + HelpRequiringPassphrase(pwallet) +
+            "\nArguments:\n"
+            "1. \"flushkeypool\"       (boolean, optional, default=true) Whether to flush old unused addresses from the keypool.\n"
```
Text should also tell user that change addresses will be removed(or not)

achow101 replied:
Done. Also included some more text explaining hd chain split stuff.

Empact commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,73 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    CPubKey master_pub_key;
+    if (!request.params[1].isNull()) {
+        CBitcoinSecret secret;
+        if (!secret.SetString(request.params[1].get_str())) throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
```
Nit: IMO this inline throw style makes the code less readable, and it's not used elsewhere in the file.

achow101 replied:
Fixed this and other places that have similar inline ifs that are hard to read.

jonasschnelli commented:
I think this can be useful, though the following potential risks may be harmful in certain scenarios:
* Missing key rotation: generating a new hd seed may imply for novice users that this protects from a compromised seed/masterkey.
* seed option: if someone uses the optional seed parameter, it's possible that the child keys have already been used to send funds to. Eventually a rescan or a UTXO-set scan should be considered?
* "Upgrading" existing non-hd wallets: leads to a mix of HD non HD-keys. Hard to "untangle" the non HD keys in case one wants to forward all funds to the new hd keyspace.

achow101 commented:
> Missing key rotation: generating a new hd seed may imply for novice users that this protects from a compromised seed/masterkey.

The help text could be updated to indicate that only coins sent to newly generated addresses will have keys generated with the new seed.

> seed option: if someone uses the optional seed parameter, it's possible that the child keys have already been used to send funds to. Eventually a rescan or a UTXO-set scan should be considered?

I could add a rescan option that is disabled by default. It could only be enabled if explicitly set and if the keypool is regenerated.

> "Upgrading" existing non-hd wallets: leads to a mix of HD non HD-keys. Hard to "untangle" the non HD keys in case one wants to forward all funds to the new hd keyspace.

At the time of upgrade, one could send all coins to a new address.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() > 2) {
```
Could have a test for extra parameters. See [example](https://github.com/bitcoin/bitcoin/pull/12083/files#diff-695d736c7d2a38a42ae900546aed52f8).

achow101 replied:
Done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    }
+
+    CPubKey master_pub_key;
+    if (!request.params[1].isNull()) {
```
Maybe:
```cpp
if (request.params[1].isNull()) {
    master_pub_key = pwallet->GenerateNewHDMasterKey();
} else {
```

achow101 replied:
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    if (!request.params[1].isNull()) {
+        CBitcoinSecret secret;
+        if (!secret.SetString(request.params[1].get_str())) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+
+        CKey key = secret.GetKey(), key2;
+        if (!key.IsValid()) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Private key outside allowed range");
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+
+    bool flush_key_pool = true;
+    if (!request.params[0].isNull()) {
+        flush_key_pool = request.params[0].get_bool();
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+        // check that we don't already have this key, compressed or otherwise
+        key2.Set(key.begin(), key.end(), !key.IsCompressed());
+        if (pwallet->HaveKey(key.GetPubKey().GetID()) || pwallet->HaveKey(key2.GetPubKey().GetID())) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Already have this key (either as an HD seed or as a loose private key)");
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    // Upgrade the wallet to version FEATURE_HD_SPLIT if it isn't aleady
+    pwallet->SetMinVersion(FEATURE_HD_SPLIT);
+
+    if (!pwallet->SetHDMasterKey(master_pub_key)) {
```
Removed the if block

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    }
+    if (flush_key_pool) pwallet->NewKeyPool();
+
+    return true;
```
Changed to `NullUniValue`

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            "\nNote that you will need to MAKE A NEW BACKUP of your wallet after setting the HD wallet seed.\n"
+            + HelpRequiringPassphrase(pwallet) +
+            "\nArguments:\n"
+            "1. \"flushkeypool\"       (boolean, optional, default=true) Whether to flush old unused addresses, including change addresses, from the keypool.\n"
```
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3788,6 +3788,81 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    if (!pwallet->SetHDMasterKey(master_pub_key)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Storing master key failed");
+    }
+    if (flush_key_pool) pwallet->NewKeyPool();
```
That's fine. It only returns false when the wallet is locked, and that is caught earlier with `EnsureWalletIsUnlocked`

achow101 commented:
Had to rebase this.

instagibbs commented on test/functional/wallet_hd.py:
```diff
@@ -118,5 +119,38 @@ def run_test (self):
+        new_masterkeyid = self.nodes[1].getwalletinfo()['hdmasterkeyid']
+        assert orig_masterkeyid != new_masterkeyid
+        addr = self.nodes[1].getnewaddress()
+        assert_equal(self.nodes[1].getaddressinfo(addr)['hdkeypath'], 'm/0\'/0\'/0\'') # Make sure the new address is the first from the keypool
```
nit: also check it's not the same key(not just path)

achow101 replied:
Done

instagibbs commented on test/functional/wallet_hd.py:
```diff
@@ -118,5 +119,38 @@ def run_test (self):
+        assert_equal(self.nodes[1].getaddressinfo(addr)['hdkeypath'], 'm/0\'/0\'/0\'') # Make sure the new address continues previous keypool
+
+        # Sethdseed parameter validity
+        assert_raises_rpc_error(-1, 'sethdseed', self.nodes[0].sethdseed, False, new_seed, 0)
```
is this for invalid named arguments? Without running this I'm not sure what it's doing

achow101 replied:
This just tests that an error is thrown when there are too many arguments

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3815,6 +3815,73 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            "                             If true, the next address from getnewaddress and change address from getrawchangeaddress will be from this new seed.\n"
+            "                             If false, addresses (including change addresses if the wallet already had HD Chain Split enabled) from the existing \n"
+            "                             keypool will be used until it has been depleted.\n"
+            "2. \"seed\"               (string, optional) The WIF private key to use as the new HD seed;\n"
```
I think it stands to be noted that Core software will never(?) return this value on a currently running wallet. You can get address' keys, but not the seed itself. I can easily see users getting confused and thinking that backing up some address key will save their wallet in the future.

achow101 replied:
Done

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3815,6 +3815,73 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key");
+        }
+
+        // check that we don't already have this key, compressed or otherwise
```
The seed value is interpreted and saved as private key. We don't want to have a seed where we already have that seed value as a key in the wallet. The only way to check whether we have a given private key is to derive its public key, get the id of that public key, and look it up. So to ensure that we don't have the seed value already, we need to derive both compressed and uncompressed public keys, get their ids, and look them up.

It's important to note that the seed is not actually a private key so it doesn't have a corresponding public key. We just use WIF because its a common format.

jimpo replied:
What is the problem with having a seed that is the same private key with opposite compressedness? It wouldn't have an existing entry in the the wallet file because the CKeyID is different.

If there is a good reason to protect against this, it might make sense to add a `bool HaveKey(CKey& key) const` method on `CKeyStore`.

achow101 replied:
If we use a key that already exists in the wallet but with opposite compressedness, that key could be exportable with `dumpprivkey`. But the user would not necessarily know that the key that is exported is their seed just with opposite compressedness. I don't think that we want the seed to be exportable like that, and if we do want the seed to be exported in some way, it should be via its own RPC and not through `dumpprivkey`.

jimpo replied:
OK. I'm still not convinced (I could make similar arguments about rejecting keys with the lowest bit toggled), but it doesn't really matter.

I would prefer to see the lookups moved over to a `HaveKey(CKey& key)` method on KeyStore though.

sipa replied:
@jimpo I don't understand what HaveKey(CKey) would do. There is a 1-to-1 correspondence both ways between CKeys and CPubKeys.

jimpo replied:
@sipa It would replicate the logic here that checks whether there is a CPubKey corresponding to either the provided CKey or the one with opposite compressedness. Basically, I don't think that this RPC function is the best place for that logic.

sipa replied:
@jimpo Ah, I see. My concern is adding extra unnecessary functions to the CKeyStore interface (which need to be implemented by all implementations, even though there's just one now). What would you think about having a *function* (not method) in keystore.h/cpp that does this, which takes a `const CKeyStore&` as argument?

achow101 replied:
I implemented @sipa's suggestion

TheBlueMatt commented:
At the risk of expanding the scope, can we fix -upgradewallet to work for HD as well here? The fact that upgradewallet will work to upgrade your wallet to HD-1 and then you have to call sethdseed to upgrade seems....strange.
Separately, can we either check that we're fully synced before allowing an HD master rotate or keep around old HD keys for key derivation? I'd prefer the second, but its obviously a ton more complicated, so just ensuring that we're at least synced first is likely sufficient to ensure people dont rotate and miss some payments.

achow101 committed:
Allow -upgradewallet to upgradewallets to HD

Changes the maximum upgradewallet version to the latest wallet version
number, 159900. Non-HD wallets will be upgraded to use HD derivation.
Non HD chain split wallets will be upgraded to HD chain split.

If a non-HD wallet is upgraded to HD, the keypool will be entirely
regenerated.

Since upgradewallet is effectively run during a first run, all of the
first run initial setup stuff is combined with the upgrade to HD

achow101 commented:
I changed `-upgradewallet` so it upgrades non-HD wallets to HD and HD chain split. I've updated the OP and title to indicate as such.

Unfortunately automated tests can't be done with `-upgradewallet`.

> Separately, can we either check that we're fully synced before allowing an HD master rotate 

This seems incompatible with `-upgradewallet` as an upgrade method.

> or keep around old HD keys for key derivation? I'd prefer the second, but its obviously a ton more complicated, so just ensuring that we're at least synced first is likely sufficient to ensure people dont rotate and miss some payments.

I think this should be done as a separate PR. It's quite a bit of scope creep.

TheBlueMatt commented:
> This seems incompatible with -upgradewallet as an upgrade method.

I dont care as much about it for first-upgrade (first-upgrade-to-HD is actually no different from just using your existing wallet from a backup/keypool perspective, as long as you dont flush keypool, its just a new way to generate keypool entries), but for HD master *rotation*, I think making sure we dont need any future new keys is pretty easy to add, no?

achow101 commented:
> I dont care as much about it for first-upgrade (first-upgrade-to-HD is actually no different from just using your existing wallet from a backup/keypool perspective

Right, duh.

***

I rebased this and added a check for IBD so it won't set a new seed if we are still in IBD.

achow101 committed:
Separate HaveKey function that checks whether a key is in a keystore

jimpo commented:
Thinking about this a bit more, it definitely feels odd that I can't sethdseed to a past value. If I have a seed set, then change it to a new one, then want to change it back, I'm not allowed to do that. Not a big deal I guess...

utACK modulo the linter error:
```
src/wallet/wallet.cpp:                LogPrintf("Upgrading wallet to use HD chain split");
^---- failure generated from contrib/devtools/lint-logs.sh
```

achow101 commented:
Fixed the linter I think

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -3991,37 +3991,54 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
-        if (!walletInstance->SetHDMasterKey(masterPubKey))
-            throw std::runtime_error(std::string(__func__) + ": Storing master key failed");
+        // Upgrade to HD if necessary
+        LOCK(walletInstance->cs_wallet);
```
Perhaps you wanted this `LOCK` inside the scope that follows? As is, the lock is held during the call to `SetBestChain` below.

achow101 replied:
Yes, that's what I wanted to do. Fixed.

sipa reviewed (commented):
Concept ACK

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3812,6 +3812,75 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    // Upgrade the wallet to version FEATURE_HD_SPLIT if it isn't aleady
+    pwallet->SetMinVersion(FEATURE_HD_SPLIT);
+    pwallet->SetHDMasterKey(master_pub_key);
+    if (flush_key_pool) pwallet->NewKeyPool();
```
This is not sufficient to ensure old wallets continue to operate using the existing keypool. In the case of HD split, you're gonna have an empty internal keypool and will instantly refresh your keypool, resulting in all change addresses being from the new HD seed. I dont see an obvious fix, but its definitely a major issue.

achow101 replied:
Do you think that it would be better then to not touch the version number in `sethdseed` and leave that to `-upgradewallet`? So then `sethdseed` would only work for HD wallets and retain whatever splitness is already in the wallet.

TheBlueMatt replied:
I think that would solve it for sethdseed and would certainly make this easier to review (as its otherwise the first time we're doing an upgrade at runtime), though it does limit the utility a good bit. It would not, however, solve the identical issue during -upgradewallet.

TheBlueMatt commented on src/wallet/wallet.h:
```diff
@@ -95,7 +95,7 @@ enum WalletFeature
     FEATURE_NO_DEFAULT_KEY = 159900, // Wallet without a default key written
 
-    FEATURE_LATEST = FEATURE_COMPRPUBKEY // HD is optional, use FEATURE_COMPRPUBKEY as latest version
+    FEATURE_LATEST = FEATURE_NO_DEFAULT_KEY
```
I dont think we should "upgrade" to NO_DEFAULT_KEY (unless we delete the default key, but that seems useless). Rather, we should upgrade to HD_SPLIT and then upgrade to NO_DEFAULT_KEY when we next bump FEATURE_LATEST (probably leave a comment to that effect here).

achow101 replied:
I don't see why not upgrade to NO_DEFAULT_KEY. FEATURE_LATEST is only used for new wallets and upgraded wallets; users should expect new and upgraded wallets to not be backwards compatible.

TheBlueMatt replied:
NO_DEFAULT_KEY doesnt even have a meaning in this case, I mean you're upgrading to nVersion == NO_DEFAULT_KEY but you still have a default key? Its just needless incompatibility at that point (though, I agree, not a huge deal given the user asked for that).

sipa replied:
@TheBlueMatt Does that matter? This is just the default for new wallets or explicit upgrades. I don't really care about gratuitous incompatibility when you're explicitly not caring about that.

UPDATE: especially with the new PRE_SPLIT type added afterwards this is probably inevitable.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -3987,37 +3991,54 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+            // Upgrade to HD chain split if necessary
+            if (walletInstance->CanSupportFeature(FEATURE_HD_SPLIT)) {
+                LogPrintf("Upgrading wallet to use HD chain split\n");
+                walletInstance->SetMinVersion(FEATURE_HD_SPLIT);
```
Same comment here, this implicitly will generate a new keypool on next internal key usage.

achow101 replied:
Is that a problem though? An upgrade was explicitly asked for. Would it be better to regenerate the keypool in this instance too? Either way, both the internal and external keypools will still be generated from the same seed.

TheBlueMatt replied:
No, this significantly changes the upgradewallet semantics, before you could upgrade and you'd still be able to use your backup in the same way based on the keypool size, now that is no longer true, which needs huge flashing warnings, at a minimum.

sipa replied:
Good point, @TheBlueMatt. I was forgetting about the distinction "need a new backup" and "need new software in the future".

What would you suggest as an ideal solution? `-upgradewallet` by default never making a change that requires a new backup, while having another option (`-fullupgradewallet` ?) to indicate you're aware a new backup is needed?

achow101 replied:
What about an `InitWarning` message that appears when the keypool is regenerated that a new backup will need to be made?

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -3987,37 +3991,54 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
             InitError(strprintf(_("Error creating %s: You can't create non-HD wallets with this version."), walletFile));
             return nullptr;
         }
-        walletInstance->SetMinVersion(FEATURE_NO_DEFAULT_KEY);
```
Why is it ok to remove this if (fFirstRun) SetMinVersion(NO_DEFAULT_KEY)?

achow101 replied:
`-upgradewallet` is implicitly true (and 0) on the first run. So the wallet version will be set to `FEATURE_LATEST`which is `FEATURE_NO_DEFAULT_KEY`.

TheBlueMatt replied:
Hmm, but what if I override it? Can I start my node accidentally with upgradewallet=0 have it create a wallet and end up with a borked wallet?

TheBlueMatt replied:
Still needs addressing, I think -upgradewallet=0 may break first-init after this patch.

achow101 replied:
I changed the if statement to `if (gArgs.GetBoolArg("-upgradewallet", fFirstRun) || fFirstRun)`

TheBlueMatt replied:
I still dont think this is right - I dont see where an fFirstRun wallet with upgradewallet=something non-0 in the bitcoin.conf will get its MinVersion set to at least FEATURE_NO_DEFAULT_KEY.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3812,6 +3812,75 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    }
+
+    // Upgrade the wallet to version FEATURE_HD_SPLIT if it isn't aleady
+    pwallet->SetMinVersion(FEATURE_HD_SPLIT);
```
This is explicitly mentioned in the help text.

TheBlueMatt replied:
Still probably shouldn't do it...there isnt any reason for it, is there?

achow101 replied:
I suppose there isn't.

achow101 commented:
I changed `sethdseed` to only work on HD wallets and not modify the wallet version.

sipa commented:
utACK afafc9b92dc1c501ca1ff63a5042e1925b7ac55a. I think it's fine that running with -upgradewallet defaults to NO_DEFAULT_KEY.

TheBlueMatt commented:
One option might be splitting the existing keypool between internal and external at the time of the upgrade, though that also violates previous API guarantees. The only "full" fix would require the new-key-from-pool callsites to be aware of the upgrade, though I don't think that'd be a ton of code.

On April 19, 2018 6:57:17 PM UTC, Pieter Wuille <notifications@github.com> wrote:
>sipa commented on this pull request.
>
>
>
>> +            LOCK(walletInstance->cs_wallet);
>+            bool hd_upgrade = false;
>+            if (walletInstance->CanSupportFeature(FEATURE_HD) &&
>!walletInstance->IsHDEnabled()) {
>+                LogPrintf("Upgrading wallet to HD\n");
>+                walletInstance->SetMinVersion(FEATURE_HD);
>+
>+                // generate a new master key
>+                CPubKey masterPubKey =
>walletInstance->GenerateNewHDMasterKey();
>+                if (!walletInstance->SetHDMasterKey(masterPubKey))
>+                    throw std::runtime_error(std::string(__func__) +
>": Storing master key failed");
>+                hd_upgrade = true;
>+            }
>+            // Upgrade to HD chain split if necessary
>+            if (walletInstance->CanSupportFeature(FEATURE_HD_SPLIT)) {
>+                LogPrintf("Upgrading wallet to use HD chain split\n");
>+                walletInstance->SetMinVersion(FEATURE_HD_SPLIT);
>
>Good point, @TheBlueMatt. I was forgetting about the distinction "need
>a new backup" and "need new software in the future".
>
>What would you suggest as an ideal solution? `-upgradewallet` by
>default never making a change that requires a new backup, while having
>another option (`-fullupgradewallet` ?) to indicate you're aware a new
>backup is needed?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/12560#discussion_r182850939

TheBlueMatt commented:
I don't think InitWarning is sufficient. For GUI I'd expect a pop-up asking if I'd like to continue, on bitcoind people don't read their debug.log, so unless we want to implement the "knowledge of past upgrade" in new key logic (which I think is probably doable), we probably need a different option :(.

On April 19, 2018 7:59:53 PM UTC, Andrew Chow <notifications@github.com> wrote:
>achow101 commented on this pull request.
>
>
>
>> +            LOCK(walletInstance->cs_wallet);
>+            bool hd_upgrade = false;
>+            if (walletInstance->CanSupportFeature(FEATURE_HD) &&
>!walletInstance->IsHDEnabled()) {
>+                LogPrintf("Upgrading wallet to HD\n");
>+                walletInstance->SetMinVersion(FEATURE_HD);
>+
>+                // generate a new master key
>+                CPubKey masterPubKey =
>walletInstance->GenerateNewHDMasterKey();
>+                if (!walletInstance->SetHDMasterKey(masterPubKey))
>+                    throw std::runtime_error(std::string(__func__) +
>": Storing master key failed");
>+                hd_upgrade = true;
>+            }
>+            // Upgrade to HD chain split if necessary
>+            if (walletInstance->CanSupportFeature(FEATURE_HD_SPLIT)) {
>+                LogPrintf("Upgrading wallet to use HD chain split\n");
>+                walletInstance->SetMinVersion(FEATURE_HD_SPLIT);
>
>What about an `InitWarning` message that appears when the keypool is
>regenerated that a new backup will need to be made?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/12560#discussion_r182867406

achow101 commented:
> For GUI I'd expect a pop-up asking if I'd like to continue

There's `UIInterface.ThreadSafeQuestion` which can be used for that.

> on bitcoind people don't read their debug.log

`InitError`, `InitWarning`, and `ThreadSafeQuestion` all print to stderr for bitcoind (unless `-daemon`). Also a recent change makes `-printtoconsole` default when using without `-daemon`.

I suppose the case with `-daemon` can really only be accounted for with a separate option.

achow101 commented:
> If you get a chance to implement tracking upgrade and remaining-original-keypool stuff I have a feeling it'll be workable.

So basically use the original keypool until it runs out, and then new keys are generated from the seed? So the callers (particularly getting change keys) will need to know whether it should use external keypool if it was upgraded but not run out yet?

I assume that this will still require warning that a new backup is required.

TheBlueMatt commented:
Yes, but that doesn't need a new backup any more than using up your current keypool does. The real issue is I believe we'd need a version bump for it (and, thus, you'd need to upgrade to the new version and wouldn't be able to upgrade to HD/HD_SPLIT/NO_DEFAULT_KEY but would have to skip them and upgrade to the new version).

On April 19, 2018 8:30:18 PM UTC, Andrew Chow <notifications@github.com> wrote:
>> If you get a chance to implement tracking upgrade and
>remaining-original-keypool stuff I have a feeling it'll be workable.
>
>So basically use the original keypool until it runs out, and then new
>keys are generated from the seed? So the callers (particularly getting
>change keys) will need to know whether it should use external keypool
>if it was upgraded but not run out yet?
>
>I assume that this will still require warning that a new backup is
>required.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/12560#issuecomment-382870871

achow101 commented:
I don't think that a wallet version upgrade is necessarily required.

For upgrade from non-HD to HD, it's easy to tell when you have exhausted the original keypool: the CKeyMetadata hdMsaterKeyID for each key will change from Null to the current hdMasterKeyID.

Non chain split to chain split is harder. However it would be very easy if a new seed was used. I don't think there would be any adverse effects if the upgrade from non split to split resulted in a new seed being set, although that may be undesireable.

achow101 commented:
Actually bumping the version number and having another field is easier to do and has less weirdness with `-salvagewallet` scenarios, so I'm going to to that instead.

achow101 commented:
Hmm, `-salvagewallet` isn't actually a problem because `pool` entries aren't retained.

The question then is whether we want to generate a new seed for non-split to split upgrade or if there is some other way to detect that upgrade.

achow101 committed:
Use a keypool of presplit keys after upgrading to hd chain split

After upgrading to HD chain split, we want to continue to use keys
from the old keypool. To do this, before we generate any new keys after
upgrading, we mark all of the keypool entries as being pre-chain
split and move them to a separate pre chain split keypool. Keys are
fetched from that keypool until it is emptied. Only then are the new
internal and external keypools used.

achow101 commented:
Merge/review beg

TheBlueMatt commented:
Thanks, I do like the pre-split keypool thing. However, I do think you need to bump the version for it - otherwise you could 1) upgrade to HD, 2) downgrade one version to 0.16 (will still open your wallet), 3) end up using keys from the new keypool without realizing it. Its not exactly a huge concern, but fixing it seems to me to be ~free, so we might as well.

achow101 committed:
Bump wallet version for pre split keypool

Bump the wallet version to indicate support for the pre split keypool.
Also prevents any wallets from upgrading to versions between HD_SPLIT
and PRE_SPLIT_KEYPOOL.

achow101 commented:
I bumped the wallet version number. I've also disallowed upgrades to any version between HD_SPLIT and PRE_SPLIT_KEYPOOL.

TheBlueMatt reviewed (commented):
Looks good, will give it a full review next week!

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4115,6 +4115,77 @@ UniValue listlabels(const JSONRPCRequest& request)
+            + HelpExampleCli("sethdseed", "")
+            + HelpExampleCli("sethdseed", "false")
+            + HelpExampleCli("sethdseed", "true \"wifkey\"")
+            + HelpExampleRpc("sethdseed", "true, \"wifkey\"")
```
Is this supposed to have a comma in the string?

achow101 replied:
Yes

achow101 replied:
Really? Looks like valid to me once the full string is constructed:

`> curl --user myusername --data-binary '{"jsonrpc": "1.0", "id":"curltest", "method": "sethdseed", "params": [true, "wifkey"] }' -H 'content-type: text/plain;' http://127.0.0.1:8332/`

sipa replied:
Oops, you're right. This is very confusing, though. Can't we make `HelpExampleRpc` take a list of UniValue arguments, and have it construct the correct JSON code for it automatically (not for this PR, obviously).

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4115,6 +4115,77 @@ UniValue listlabels(const JSONRPCRequest& request)
+    }
+
+    // Do not do anything to non-HD wallets
+    if (!pwallet->IsHDEnabled()) {
```
This needs cs_wallet.

achow101 replied:
Done

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4115,6 +4115,77 @@ UniValue listlabels(const JSONRPCRequest& request)
+    if (request.params[1].isNull()) {
+        master_pub_key = pwallet->GenerateNewHDMasterKey();
+    } else {
+        CKey key = DecodeSecret(request.params[1].get_str()), key2;
```
key2 looks unused.

achow101 replied:
Removed

TheBlueMatt reviewed (commented):
Need to take appropriate action on the pre_split keys when set.*KeyPool is changed - at least the DBErrors::NEED_REWRITE handling, NewKeyPool, and ReturnKey handling looks like it should be updated.

achow101 commented:
Addressed @TheBlueMatt's comments, also rebased.

achow101 commented:
I have added checks for whether to use `set_pre_split_keypool` in places where keypools are chosen.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -4002,9 +4037,10 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
     }
 
-    if (gArgs.GetBoolArg("-upgradewallet", fFirstRun))
+    if (gArgs.GetBoolArg("-upgradewallet", fFirstRun) || fFirstRun)
```
Grr, the whole first if block *was* dead code and is no longer, making this harder to review. I think its mostly equivalent to the old stuff, but would be really nice to get the diff cleaned up here. Can we just not merge out the two if statements?

TheBlueMatt reviewed (commented):
Can you try to simplify the diff on the CreateWalletFromFile stuff, its very hard to review (and I think subtly wrong in some strange edge-cases).

achow101 commented:
@TheBlueMatt I simplified the diff by removing the combination of those two if blocks. The hd upgrading code has been moved to its own if block that enters if `-upgradewallet` is actually explicitly set.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4079,6 +4079,78 @@ static UniValue listlabels(const JSONRPCRequest& request)
+            "\nArguments:\n"
+            "1. \"newkeypool\"         (boolean, optional, default=true) Whether to flush old unused addresses, including change addresses, from the keypool and regenerate it.\n"
+            "                             If true, the next address from getnewaddress and change address from getrawchangeaddress will be from this new seed.\n"
+            "                             If false, addresses (including change addresses if the wallet already had HD Chain Split enabled) from the existing \n"
```
μ-nit: unnecessary space before `\n` at end of line.

achow101 replied:
Done

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4079,6 +4079,78 @@ static UniValue listlabels(const JSONRPCRequest& request)
+    pwallet->SetHDMasterKey(master_pub_key);
+    if (flush_key_pool) pwallet->NewKeyPool();
+
+    return NullUniValue;
```
Isn't this supposed to return true or false depending on success? From L4105 above (result).

achow101 replied:
Fixed

achow101 replied:
Actually I'll keep it as `NullUnivalue` and I updated the docs.

kallewoof replied:
Makes sense to me.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4021,14 +4057,57 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
 
+    // Upgrade to HD if explicit upgrade
+    if (gArgs.GetBoolArg("-upgradewallet", false))
+    {
```
Nit: `{` on same line as `if (..)`

achow101 replied:
Fixed

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4021,14 +4057,57 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+        // Do not upgrade versions to any version between HD_SPLIT and FEATURE_PRE_SPLIT_KEYPOOL unless already supporting HD_SPLIT
+        int nMaxVersion = walletInstance->nWalletVersion;
+        if (!walletInstance->CanSupportFeature(FEATURE_HD_SPLIT) && nMaxVersion >=FEATURE_HD_SPLIT && nMaxVersion < FEATURE_PRE_SPLIT_KEYPOOL) {
+            InitError(_("Cannot upgrade a non HD split wallet without upgrading to support pre split keypool"));
```
This error would scare me (since I'm juggling privkeys and it's not a very easy to understand message). Maybe be nice and do `[...]split keypool. To do this, run [XYZ] first.` or similar?

achow101 replied:
I edited the error to say what to do instead.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4021,14 +4057,57 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+        LOCK(walletInstance->cs_wallet);
+
+        // Do not upgrade versions to any version between HD_SPLIT and FEATURE_PRE_SPLIT_KEYPOOL unless already supporting HD_SPLIT
+        int nMaxVersion = walletInstance->nWalletVersion;
```
μ: Should try to avoid the type prefix camel case style in new code (`max_version`?).

achow101 replied:
Fixed

kallewoof reviewed (commented):
utACK 8db939a7b953335c5014147dd0d85069b606e0fb with some μ-nits

sipa commented on src/wallet/wallet.h:
```diff
@@ -141,9 +142,18 @@ class CKeyPool
                 fInternal = false;
             }
+            try {
+                READWRITE(m_pre_split);
```
You can test for this using `!(ser_action.ForRead() && s.empty())` instead of trying/catching.

achow101 replied:
I don't quite understand what you mean.

Also, I would like to keep this consistent with the block above which does the same thing.

jonasschnelli replied:
I think sipa meant to check during read if there are more bytes in the stream because this changed only appends data at the end.

sipa replied:
Indeed, what @jonasschnelli says. If it's how similar already operates we can change it later in one go, too.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4025,14 +4025,42 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
+
+            // generate a new master key
+            CPubKey masterPubKey = walletInstance->GenerateNewHDMasterKey();
+            if (!walletInstance->SetHDMasterKey(masterPubKey))
```
Fixed.

kallewoof commented:
That makes sense to me.

On Thu, May 10, 2018 at 11:31 AM, Andrew Chow <notifications@github.com>
wrote:

> *@achow101* commented on this pull request.
> ------------------------------
>
> In src/wallet/rpcwallet.cpp
> <https://github.com/bitcoin/bitcoin/pull/12560#discussion_r187226164>:
>
> > +        CKey key = DecodeSecret(request.params[1].get_str());
> +        if (!key.IsValid()) {
> +            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key");
> +        }
> +
> +        if (HaveKey(*pwallet, key)) {
> +            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Already have this key (either as an HD seed or as a loose private key)");
> +        }
> +
> +        master_pub_key = pwallet->DeriveNewMasterHDKey(key);
> +    }
> +
> +    pwallet->SetHDMasterKey(master_pub_key);
> +    if (flush_key_pool) pwallet->NewKeyPool();
> +
> +    return NullUniValue;
>
> Actually I'll keep it as NullUnivalue and I updated the docs.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/12560#discussion_r187226164>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAPRcKtxPjARR8shxzcs0FWo7WBL_D0Qks5tw6ZsgaJpZM4SVPiY>
> .
>

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4079,6 +4079,78 @@ static UniValue listlabels(const JSONRPCRequest& request)
+            "2. \"seed\"               (string, optional) The WIF private key to use as the new HD seed; if not provided a random seed will be used.\n"
+            "                             The seed value can be retrieved using the dumpwallet command. It is the private key marked hdmaster=1\n"
+            "\nResult:\n"
+            "null                    (boolean) Returns null if successful\n"
```
Just leave the result section out -- e.g. like in `setlabel` (L326).

achow101 replied:
Done

kallewoof reviewed (commented):
re-utACK ae810ff48650a07160877199f9283edfda1b3632

jonasschnelli commented on src/keystore.cpp:
```diff
@@ -195,3 +195,10 @@ CKeyID GetKeyForDestination(const CKeyStore& store, const CTxDestination& dest)
+{
+    CKey key2;
+    key2.Set(key.begin(), key.end(), !key.IsCompressed());
+    return store.HaveKey(key.GetPubKey().GetID()) || store.HaveKey(key2.GetPubKey().GetID());
```
Does this have performance impacts?
What are the downside of not checking the non-compressed (or compressed if non-compressed)?

achow101 replied:
I don't think there are performance impacts, or if there are, they are not noticeable. The reason to check for both is to avoid having a seed value which was already in the wallet before and thus could have been revealed.

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4079,6 +4079,78 @@ static UniValue listlabels(const JSONRPCRequest& request)
+    }
+
+    if (IsInitialBlockDownload()) {
+        throw JSONRPCError(RPC_CLIENT_IN_INITIAL_DOWNLOAD, "Cannot set a new HD seed while still in Initial Block Download");
```
Can you elaborate why a IBD check is necessary and why the wallet-key-state is connected to IBD?

achow101 replied:
From Matt's earlier comments:

> Separately, can we either check that we're fully synced before allowing an HD master rotate or keep around old HD keys for key derivation? I'd prefer the second, but its obviously a ton more complicated, so just ensuring that we're at least synced first is likely sufficient to ensure people dont rotate and miss some payments.

jonasschnelli commented:
Concept ACK, will test more later.

sipa reviewed (commented):
utACK a8da482a8bc87ff26194612727d4a7b86b2fb60d

MarcoFalke commented:
@jonasschnelli You added this to the 15.2 milestone without the "needs backport" label. Was this done in mistake? I think a backport should be discussed shortly on irc

laanwj commented:
utACK a8da482a8bc87ff26194612727d4a7b86b2fb60d
(hope this works, the github unicorn is preventing me from going to this PR
while logged in)
